src/net.cpp

```diff
@@ -628,6 +628,135 @@ int CNode::GetSendVersion() const
+        }
+        for (uint256 wtxid : m_recon_state->local_set_snapshot) {
+            short_ids.push_back(ComputeShortID(wtxid));
+        }
+
+        std::vector<uint32_t> bisec_short_ids;
+        for (uint32_t short_id : short_ids) {
+            if (bisection_chunk == BISECTION_LOW) {
+                if (short_id <= BISECTION_MEDIAN) {
+                    bisec_short_ids.push_back(short_id);
+                }
+            } else {
+                if (short_id > BISECTION_MEDIAN) {
+                    bisec_short_ids.push_back(short_id);
+                }
+            }
+        }
+        short_ids = bisec_short_ids;
+        // For bisection, use capacity used in the initial reconciliation.
+        capacity = m_recon_state->capacity_snapshot;
+    } else {
+        for (uint256 wtxid : m_recon_state->local_set) {
+            short_ids.push_back(ComputeShortID(wtxid));
+        }
+
+        if (capacity == 0) {
+            // Estimate locally, as requested by the calling function.
+            int set_size_diff = m_recon_state->local_set.size() - m_recon_state->remote_set_size;
+            double q = use_own_q ? m_recon_state->local_q : m_recon_state->remote_q;
+            capacity = 1 + q * fmin(m_recon_state->local_set.size(), m_recon_state->remote_set_size) +
+                       abs(set_size_diff);
+        }
+        capacity = std::min(capacity, MAX_SKETCH_CAPACITY);
+        // If bisection is required, we will use this capacity from the initial reconciliation.
+        m_recon_state->capacity_snapshot = capacity;
+    }
+    if (short_ids.size() == 0) return nullptr;
+    minisketch* sketch = minisketch_create(RECON_FIELD_SIZE, 0, capacity);
+    for (uint32_t short_id : short_ids) {
+        minisketch_add_uint64(sketch, short_id);
+    }
+    return sketch;
+}
+
+std::vector<uint256> CNode::GetRelevantIDsFromShortIDs(uint64_t diff[], uint8_t diff_size, std::vector<uint32_t>& local_missing, CConnman* connman)
+{
+    std::vector<uint256> remote_missing;
+    for (int i = 0; i < diff_size; ++i) {
+        auto local_tx = m_recon_state->local_short_id_mapping.find(diff[i]);
+        if (local_tx != m_recon_state->local_short_id_mapping.end()) {
+            remote_missing.push_back(local_tx->second);
+        } else {
+            local_missing.push_back(diff[i]);
+        }
+    }
+    return remote_missing;
+}
+
+std::vector<uint256> CNode::GetWTXIDsFromShortIDs(const std::vector<uint32_t> remote_missing_short_ids, CConnman* connman)
+{
+    std::vector<uint256> remote_missing;
+    for (size_t i = 0; i < remote_missing_short_ids.size(); ++i) {
+        auto local_tx = m_recon_state->local_short_id_mapping.find(remote_missing_short_ids[i]);
+        if (local_tx == m_recon_state->local_short_id_mapping.end()) {
+            LogPrint(BCLog::NET, "Unknown transaction requested by short id=%d by peer=%d\n", remote_missing_short_ids[i], GetId());
+        }
+        remote_missing.push_back(local_tx->second);
```

Comment:
I think that this is UB (if no wtxid for a shortid is found, there is a LogPrint, but the item is still dereferenced).
While this should explain the strange wtxids with lots of 0s, it doesn't explain the why no mapping is found to begin with.

Reply:
I found an issue, will commit the fix later today!
Basically, sometimes reconcil. difference finding gives a false positive: it looks like a success, but some difference elements are missing.
This is just a natural property of math in minisketch. I'll submit a workaround.

Reply:
Still not sure why it would be UB here? It just accesses the existing array element..

Reply:
You can't dereference local_tx when it is .end().

Reply:
I may be off, but if `local_tx == m_recon_state->local_short_id_mapping.end()`, `local_tx->second` is still accessed after the LogPrint - isn't that illegal?

Reply:
Oh, i see you are definitely right, it's a bug!